File: packages/api/src/domain/household.ts
import { isCurrencyCode, type CurrencyCode } from '@trata/money'
import { asDateTimeString, asNonEmptyString, asString, isRecord } from '../lib/normalize'

export type HouseholdRole = 'owner' | 'member'

export interface HouseholdMember {
  userId: string
  email: string
  /** Display name; null when never set (consumers fall back to email). */
  displayName: string | null
  role: HouseholdRole
  joinedAt: string
}

export interface Household {
  id: string
  createdAt: string
  /** Owner-set display name; null = never set (consumers derive a label). */
  name: string | null
  /**
   * Base currency: the conversion target for household-wide presentation.
   * Owner-editable; changing it never rewrites stored amounts.
   */
  currency: CurrencyCode
  members: HouseholdMember[]
}

/** Invitation lifecycle state as reported by the owner-side listing. */
export type HouseholdInvitationStatus = 'pending' | 'accepted' | 'revoked' | 'expired'

export interface HouseholdInvitation {
  id: string
  email: string
  status: HouseholdInvitationStatus
  createdAt: string
  expiresAt: string
  acceptedAt: string | null
  revokedAt: string | null
}

/** Acceptor-side preview of an invitation (matching-email accounts only). */
export interface HouseholdInvitationPreview {
  householdName: string | null
  membersCount: number
  inviterEmail: string
  inviterDisplayName: string | null
  expiresAt: string
}

export interface HouseholdCode {
  code: string
  createdAt: string
}

const isHouseholdRole = (value: unknown): value is HouseholdRole =>
  value === 'owner' || value === 'member'

const isInvitationStatus = (value: unknown): value is HouseholdInvitationStatus =>
  value === 'pending' ||
  value === 'accepted' ||
  value === 'revoked' ||
  value === 'expired'

// Nullable-but-optional field: the Go server omits nil pointers (omitempty),
// so "absent" and JSON null are the same thing here - both mean "not set".
const asNullableString = (value: unknown): string | null =>
  value === null || value === undefined ? null : asString(value)

const asNullableDateTime = (value: unknown): string | null =>
  value === null || value === undefined ? null : asDateTimeString(value)

export const normalizeHouseholdMember = (value: unknown): HouseholdMember | null => {
  if (!isRecord(value)) {
    return null
  }

  const userId = asNonEmptyString(value.userId)
  const email = asNonEmptyString(value.email)
  const displayName = value.displayName == null ? null : asString(value.displayName)
  const role = isHouseholdRole(value.role) ? value.role : null
  const joinedAt = asDateTimeString(value.joinedAt)

  if (!userId || !email || !role || !joinedAt) {
    return null
  }

  return { userId, email, displayName, role, joinedAt }
}

export const normalizeHousehold = (value: unknown): Household | null => {
  if (!isRecord(value) || !Array.isArray(value.members)) {
    return null
  }

  const id = asNonEmptyString(value.id)
  const createdAt = asDateTimeString(value.createdAt)
  const name = asNullableString(value.name)
  const currency = isCurrencyCode(value.currency) ? value.currency : null
  if (!id || !createdAt || !currency) {
    return null
  }

  const members: HouseholdMember[] = []
  for (const raw of value.members) {
    const member = normalizeHouseholdMember(raw)
    if (!member) {
      return null
    }
    members.push(member)
  }

  return { id, createdAt, name, currency, members }
}

export const normalizeHouseholdInvitation = (value: unknown): HouseholdInvitation | null => {
  if (!isRecord(value)) {
    return null
  }

  const id = asNonEmptyString(value.id)
  const email = asNonEmptyString(value.email)
  const status = isInvitationStatus(value.status) ? value.status : null
  const createdAt = asDateTimeString(value.createdAt)
  const expiresAt = asDateTimeString(value.expiresAt)
  const acceptedAt = asNullableDateTime(value.acceptedAt)
  const revokedAt = asNullableDateTime(value.revokedAt)
  if (!id || !email || !status || !createdAt || !expiresAt) {
    return null
  }

  return { id, email, status, createdAt, expiresAt, acceptedAt, revokedAt }
}

export const normalizeHouseholdInvitationPreview = (
  value: unknown,
): HouseholdInvitationPreview | null => {
  if (!isRecord(value)) {
    return null
  }

  const householdName = asNullableString(value.householdName)
  const inviterEmail = asNonEmptyString(value.inviterEmail)
  const inviterDisplayName = asNullableString(value.inviterDisplayName)
  const expiresAt = asDateTimeString(value.expiresAt)
  const membersCount =
    typeof value.membersCount === 'number' && Number.isSafeInteger(value.membersCount)
      ? value.membersCount
      : null
  if (!inviterEmail || !expiresAt || membersCount === null || membersCount < 1) {
    return null
  }

  return { householdName, membersCount, inviterEmail, inviterDisplayName, expiresAt }
}

export const normalizeHouseholdCode = (value: unknown): HouseholdCode | null => {
  if (!isRecord(value)) {
    return null
  }

  const code = asNonEmptyString(value.code)
  const createdAt = asDateTimeString(value.createdAt)
  if (!code || !createdAt) {
    return null
  }

  return { code, createdAt }
}

File: backend/internal/domain/household.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// HouseholdRole is a member's role in a household. Roles are stored in v1 but
// carry no behavioral split yet (ADR-0002); both roles see and modify all of
// the household's data equally.
type HouseholdRole string

const (
	HouseholdRoleOwner  HouseholdRole = "owner"
	HouseholdRoleMember HouseholdRole = "member"
)

// Household is the shared data space: every account, category, transaction,
// debtor, debt operation, and planned payment belongs to exactly one
// household, and members access it equally. Name is the optional owner-set
// display name (nil = never set; consumers derive a label from the owner).
// Currency is the base currency - the presentation conversion target; editing
// it never rewrites stored amounts.
type Household struct {
	ID        uuid.UUID
	CreatedAt time.Time
	Name      *string
	Currency  string
	Members   []HouseholdMember
}

// HouseholdMember is one user's membership in a household, joined with the
// profile fields the member listing exposes. DisplayName is nil when never
// set (consumers fall back to Email).
type HouseholdMember struct {
	UserID      uuid.UUID
	Email       string
	DisplayName *string
	Role        HouseholdRole
	JoinedAt    time.Time
}

// Membership is a user's single (v1) household membership row - the auth
// middleware's household resolution input.
type Membership struct {
	HouseholdID uuid.UUID
	UserID      uuid.UUID
	Role        HouseholdRole
	JoinedAt    time.Time
}

// HouseholdInvitationStatus is the owner-side listing state of an
// invitation; Expired is derived at read time (pending + past ExpiresAt).
type HouseholdInvitationStatus string

const (
	HouseholdInvitationPending  HouseholdInvitationStatus = "pending"
	HouseholdInvitationAccepted HouseholdInvitationStatus = "accepted"
	HouseholdInvitationRevoked  HouseholdInvitationStatus = "revoked"
	HouseholdInvitationExpired  HouseholdInvitationStatus = "expired"
)

// HouseholdInvitation is an owner-issued email invitation. The accept token
// is single-use, delivered by email, and refreshed (not duplicated) when the
// owner re-invites the same pending email.
type HouseholdInvitation struct {
	ID          uuid.UUID
	HouseholdID uuid.UUID
	Email       string
	Token       uuid.UUID
	CreatedBy   uuid.UUID
	CreatedAt   time.Time
	ExpiresAt   time.Time
	AcceptedAt  *time.Time
	RevokedAt   *time.Time
}

// Status derives the listing state: accepted/revoked win, then expiry.
func (i HouseholdInvitation) Status(now time.Time) HouseholdInvitationStatus {
	switch {
	case i.AcceptedAt != nil:
		return HouseholdInvitationAccepted
	case i.RevokedAt != nil:
		return HouseholdInvitationRevoked
	case now.After(i.ExpiresAt):
		return HouseholdInvitationExpired
	default:
		return HouseholdInvitationPending
	}
}

// HouseholdInvitationPreview is the acceptor-side view of a pending
// invitation: what the accept screen shows before the data choice.
type HouseholdInvitationPreview struct {
	HouseholdName      *string
	MembersCount       int
	InviterEmail       string
	InviterDisplayName *string
	ExpiresAt          time.Time
}

// HouseholdCode is the household's multi-use join code (family fallback; binds
// no identity). Exactly one code row per household: rotate replaces, revoke
// deactivates.
type HouseholdCode struct {
	HouseholdID uuid.UUID
	Code        string
	CreatedAt   time.Time
	RevokedAt   *time.Time
}
